app/store/actions/healthcareAuthorities/getHealthcareAuthoritiesAction.ts

```diff
@@ -0,0 +1,55 @@
+import { createAction } from '@reduxjs/toolkit';
+import Yaml from 'js-yaml';
+import { Dispatch } from 'redux';
+
+import { AUTHORITIES_LIST_URL_MVP1 } from '../../../constants/authorities';
+import { AppThunk } from '../../types';
+
+const GET_HEALTHCARE_AUTHORITIES_STARTED = 'GET_HEALTHCARE_AUTHORITIES_STARTED';
+const GET_HEALTHCARE_AUTHORITIES_SUCCESS = 'GET_HEALTHCARE_AUTHORITIES_SUCCESS';
+const GET_HEALTHCARE_AUTHORITIES_FAILURE = 'GET_HEALTHCARE_AUTHORITIES_FAILURE';
+
+const getHealthcareAuthorities_started = createAction(
+  GET_HEALTHCARE_AUTHORITIES_STARTED,
+);
+const getHealthcareAuthorities_success = createAction(
+  GET_HEALTHCARE_AUTHORITIES_SUCCESS,
+  ({ healthcareAuthorities }) => ({ payload: { healthcareAuthorities } }),
+);
+const getHealthcareAuthorities_failure = createAction(
+  GET_HEALTHCARE_AUTHORITIES_FAILURE,
+  ({ error }) => ({ payload: {}, meta: { error } }),
+);
+
+const getHealthcareAuthoritiesApi = async () => {
+  const yamlString = await fetch(AUTHORITIES_LIST_URL_MVP1).then((res) =>
+    res.text(),
+  );
+  const { authorities } = Yaml.safeLoad(yamlString);
+  if (!Array.isArray(authorities)) {
+    throw new Error('authorities yaml did not return an array of authorities');
```

Comment:
Totally agreeing on having a dedicated API folder for the app. Typically I like segmenting 
```
app/
  api/
  store/
  components/
  ...
```
I'm not a fan of how I set up the APIs contracts are being determined in the components right now, I'll make an API folder to improve that. 

I think creating an interface for error types is a good idea, but I disagree on extending that to encapsulate all responses -- I think the cleanest way for an async api abstraction to occur is for it to have the form of:

(very rough pseudo:)
```
const exampleApi = async (): Foo => {
   try {
     const foo = await fetch(...);
     if (!validateFoo(foo)) { 
       throw ValidationError
     }
   catch (e) {
     throw createTypeSafeError(e);
}
```

Essentially, following the Axios model of either APIs resolve successfully or throw with an error. I like this because it makes chaining error handling in API much easier (Like in the try in the thunk). When errors resolve into successful response bodies, though, I find it harder to have consumers implement error chaining.

Reply:
> I think creating an interface for error types is a good idea, but I disagree on extending that to encapsulate all responses -- I think the cleanest way for an async api abstraction to occur is for it to have the form of:

To be clear, I wasnt suggesting creating an error type that covers **all** responses.  usually I see an error type for each response, but, imo, it depends on the api and what the intent of the code is to determine on what makes sense.

I do think it makes sense to have a generic type for network responses in general though.

> When errors resolve into successful response bodies, though, I find it harder to have consumers implement error chaining.

Hmm, I am confused on what you mean by errors resolving into successful response bodies here. do you mean network error? or the type of error when you get a successful response but the data contains an 'error' field? or the type of error where the data was in a form that wasn't expected? And how are they resolving into successes?

---

Oops, I see the source of the confusion now. totally my bad for being sloppy in the comment above. What i meant to say was that it strikes me as odd to be throwing errors for the data validations and that it would be nice to have a NetworkResponse type.

what i meant to write in the code snippet was something like:

```javascript
const getHealthcareAuthoritiesApi = async (): Promise<NetworkResponse<HealthcareAuthority[], HealtcareAuthorityError>> => {
  try {
  const response = await fetch(url).then((res) =>
    res.text(),
  );
  const { authorities } = Yaml.safeLoad(response);
  if (!Array.isArray(authorities)) {
    return {
       kind: 'failure',
       error: 'NotANetworkError'
  } else {
  return {
      kind: 'success',
      body: authorities
    }
  }
  } catch (e) {
  return { kind: 'failure', error: 'NetworkError'}
}
};
```

I forgot the try/catch and the Promise in the type signature.

so you wouldnt be able to chain on with a .catch(), but you could still .then(), I guess I dont see validation errors as exceptions, so handling them in the normal control flow seems correct to me.

Either way, this thread is getting a bit long, and your implementation totally works; so if you prefer that, you should totally go for it!